### PR TITLE
Feature/vmp 4 add utils 2

### PR DESCRIPTION
**Pull request by AdamWiktor**

### CODE REVIEW

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+import static java.util.stream.Collectors.toList;
+import static org.opentripplanner.graph_builder.linking.LinkingGeoTools.RADIUS_DEG;
+
+/**
+ * Finds all closest edges in graph that given vertex could be linked to
+ */
+public class EdgesToLinkFinder {
+
+    /**
+     * if there are two ways and the distances to them differ by less than this value, we link to both of them
```
Comment (MironSz) [inline src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java]: Wiem, że głupota ale zaczynaj komentarze z wielkiej litery :D

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+    private TIntDoubleMap getDistances(List<StreetEdge> candidateEdges, Vertex vertex) {
+        TIntDoubleMap distances = new TIntDoubleHashMap();
+        candidateEdges.forEach(e -> distances.put(e.getId(), linkingGeoTools.distance(vertex, e)));
+        return distances;
+    }
+
+    private void sortCandidateEdges(List<StreetEdge> candidateEdges, TIntDoubleMap distances) {
+        // Sort the list.
+        candidateEdges.sort((o1, o2) -> {
```
Comment (MironSz) [inline src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java]: Nie mam pojęcia jak z wydajnością ale lepiej najpierw przemapować na (krawędź, odległość) a później sortować (wiem, że to hashmapa).

**Review by MironSz — APPROVED** [review]
Zostawiłem dwie mini uwagi, jeśli uznasz je za istotne to zmień ale ogólnie to daje okejkę. Duże kudosy za dużo testów

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java (hunk 1 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java]: Zrobione 👍

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java (hunk 2 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java]: Zrobione 👍

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+/**
+ * Finds all closest edges in graph that given vertex could be linked to
+ */
+public class EdgesToLinkFinder {
+
+    /**
+     * If there are two ways and the distances to them differ by less than this value, we link to both of them
+     */
+    public static final double DUPLICATE_WAY_EPSILON_DEGREES = SphericalDistanceLibrary.metersToDegrees(0.001);
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:30]: tak się zastanawiam, czy to jest dysnatns jednego milimetra?

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+        // Then we link to everything that is within DUPLICATE_WAY_EPSILON_METERS of of the best distance
+        // so that we capture back edges and duplicate ways.
+        return idx.query(env).stream()
+                .filter(StreetEdge.class::isInstance)
+                .map(StreetEdge.class::cast)
+                // note: not filtering by radius here as distance calculation is expensive
+                // we do that below.
+                .filter(edge -> edge.canTraverse(traverseModeSet) &&
+                        // only link to edges still in the graph.
+                        edge.getToVertex().getIncoming().contains(edge))
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:76]: rozumiem, że to skopiowane. Ale możesz wyjaśnić co to robi? Bo to jakaś tautologia chyba?

File: src/main/java/org/opentripplanner/graph_builder/linking/Linker.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+import org.opentripplanner.routing.location.TemporaryStreetLocation;
+import org.opentripplanner.routing.vertextype.StreetVertex;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Tries to link vertices to all closest edges in graph
+ */
+public class Linker {
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/Linker.java]: Nie podoba mi się nazwa.
Może `VertexLinker`?

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    public List<StreetEdge> findEdgesToLink(Vertex vertex, TraverseMode traverseMode) {
+        Envelope env = linkingGeoTools.createEnvelope(vertex);
+
+        List<StreetEdge> candidateEdges = getCandidateEdges(env, traverseMode);
+
+        List<Map.Entry<StreetEdge, Double>> sortedCandidateEdges = getSortedCandidateEdges(candidateEdges, vertex);
+
+        if (sortedCandidateEdges.isEmpty() || sortedCandidateEdges.get(0).getValue() > RADIUS_DEG) {
+            return emptyList();
+        }
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:53]: Mam wrażenie, że to należy do `getBestEdges(sortedCandidateEdges)`

Comment (Kamyki) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:53]: Ale mogę się mylić

File: src/main/java/org/opentripplanner/graph_builder/linking/Linker.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Permanently link this vertex into the graph
+     */
+    public boolean linkPermanently(Vertex vertex, TraverseMode traverseMode) {
+        List<StreetEdge> streetEdges = edgesToLinkFinder.findEdgesToLink(vertex, traverseMode);
+        streetEdges.forEach(edge -> linkPermanently(vertex, edge));
+        return !streetEdges.isEmpty();
+    }
+
+    private void linkTemporarily(TemporaryStreetLocation vertex, StreetEdge edge, RoutingRequest options) {
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/Linker.java]: trochę zonk, że ta sama nazwa co publiczna metoda. Mozę `linkTemporarilyToEdge`

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        this.edgeFactory = edgeFactory;
+        this.splitter = splitter;
+        this.edgesMaker = edgesMaker;
+        this.linkingGeoTools = linkingGeoTools;
+    }
+
+    public void setAddExtraEdgesToAreas(boolean addExtraEdgesToAreas) {
+        this.addExtraEdgesToAreas = addExtraEdgesToAreas;
+    }
+
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java:50]: Stateful?
Może bezpieczniej dodać ten parametr do wszystkich metod?

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+import org.opentripplanner.routing.vertextype.StreetVertex;
+import org.opentripplanner.routing.vertextype.TransitStop;
+import org.opentripplanner.util.I18NString;
+import org.opentripplanner.util.LocalizedString;
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Links vertex to edge in graph at a given faction
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java]: faction?

File: src/main/java/org/opentripplanner/graph_builder/linking/Linker.java (hunk 5 above)
Comment (Kamyki) [reply src/main/java/org/opentripplanner/graph_builder/linking/Linker.java]: I dodałbym w docstringu, że ta klaasa to entrypoint, bo z tych lakonicznych oposów wynika, że co druga dodaje wierzchołek/krawędź i splituje

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+        for (AreaEdge areaEdge : area.getEdges()) {
+            if (!vertices.contains(areaEdge.getToVertex())) vertices.add(areaEdge.getToVertex());
+            if (!vertices.contains(areaEdge.getFromVertex())) vertices.add(areaEdge.getFromVertex());
+        }
+
+        for (Vertex vertex : vertices) {
+            if (vertex instanceof StreetVertex && !vertex.equals(splitterVertex)) {
+                LineString line = linkingGeoTools.createLineString(splitterVertex, vertex);
+                double length = SphericalDistanceLibrary.distance(splitterVertex.getCoordinate(), vertex.getCoordinate());
+                I18NString name = new LocalizedString("", new OSMWithTags());
```
Comment (Kamyki) [inline src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java:88]: pusty string?

File: src/test/java/org/opentripplanner/graph_builder/linking/ToEdgeLinkerTest.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+
+        // then
+        verify(streetSplitter, times(1)).splitTemporarily(edge, ll, temporaryStreetLocation.isEndVertex());
+        verify(edgesMaker, times(1)).makeTemporaryEdges(eq(temporaryStreetLocation), any());
+        verifyNoMoreInteractions(streetSplitter, edgesMaker);
+        verifyZeroInteractions(linkingGeoTools);
+    }
+
+    @Test
+    public void shouldLinkVertexOnEdgePermanently() {
```
Comment (Kamyki) [inline src/test/java/org/opentripplanner/graph_builder/linking/ToEdgeLinkerTest.java:63]: Nie testujesz AreaEdge case

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java (hunk 3 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:30]: Yup, nie wiem kiedy tak mała różnica mogłaby wystąpić, ale podejrzewam, że może to mieć związek ze stratną reprezentacją liczb rzeczywistych przez komputer.

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java (hunk 4 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:76]: Jeśli usuniemy jakąś krawędź z grafu, to tak na prawdę usuwamy ją z `outgoing` jej początkowego wierzchłka oraz z `incoming` jej końcowego wierzchołka. Fizycznie obiekt `StreetEdge` będzie istniał nadal i będzie się ciągle znajdował w spatial index. Podejrzewam, że nie usuwamy krawędzi z indeksu z powodów wydajnościowych.

File: src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java (hunk 6 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/EdgesToLinkFinder.java:53]: Tutaj pójdzie jeszcze jeden refactor w następnym PR.

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java (hunk 9 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java]: fraction 👍

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java (hunk 10 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java:88]: wydaje mi się, że te krawędzie nie potrzebują mieć nazw - tak było to tworzone wcześniej

File: src/test/java/org/opentripplanner/graph_builder/linking/ToEdgeLinkerTest.java (hunk 11 above)
Comment (AdamWiktor) [reply src/test/java/org/opentripplanner/graph_builder/linking/ToEdgeLinkerTest.java:63]: To dlatego, że nie wiem jak działa `AreaEdge`, więc po prostu przekopiowałem stary kod xD

File: src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java (hunk 8 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/ToEdgeLinker.java:50]: Przerobię to w następnym PR, bo to może być `final` i ustawiane w konstruktorze

File: src/main/java/org/opentripplanner/graph_builder/linking/Linker.java (hunk 7 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/Linker.java]: done

File: src/main/java/org/opentripplanner/graph_builder/linking/Linker.java (hunk 5 above)
Comment (AdamWiktor) [reply src/main/java/org/opentripplanner/graph_builder/linking/Linker.java]: zrobione 👍
